src/lib/communication/DY.Lib.Communication.Core.Invariants.fst

```diff
@@ -17,46 +17,52 @@ open DY.Lib.Communication.Core
 
 (*** PkEnc Predicates ***)
 
-val pke_crypto_predicates_communication_layer: {|cusages:crypto_usages|} -> pke_crypto_predicate
-let pke_crypto_predicates_communication_layer #cusages = {
+#push-options "--ifuel 1"
+val pke_crypto_predicates_communication_layer: {|cusages:crypto_usages|} -> a:Type0 -> {|comm_layer_core_config a|} -> pke_crypto_predicate
+let pke_crypto_predicates_communication_layer #cusages a #config  = {
   pred = (fun tr sk_usage pk msg ->
     (exists sender receiver.
-      sk_usage == long_term_key_type_to_usage (LongTermPkeKey comm_layer_pkenc_tag)  receiver /\
+      sk_usage == long_term_key_type_to_usage (LongTermPkeKey (comm_layer_pkenc_tag a))  receiver /\
       (get_label tr msg) `can_flow tr` (comm_label sender receiver) /\
-      event_triggered tr sender (CommConfSendMsg sender receiver msg)
+      (match parse a msg with
+      | None -> False
+      | Some msg_parsed -> event_triggered tr sender (CommConfSendMsg sender receiver msg_parsed <: communication_core_event a))
+      
```

Comment:
`DY.Lib.Comparse.DYUtils.parse_and_pred msg (fun msg_parsed -> event_triggered tr sender (CommConfSendMsg sender receiver msg_parsed)` does the same thing as this. It's not that much cleaner, though, since the predicate itself doesn't have a nice name.